File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_11.java
package org.twyszomirski.aoc.solutions;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;


import static java.io.IO.println;

public class Day_11 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_11.txt"));
        println("======== Day 11 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var stones = Arrays.stream(lines.getFirst().split(" ")).mapToLong(Long::valueOf).boxed().toList();
        var sum = stones.stream().mapToLong(s -> blinkStone(s, 25, new HashMap<>())).sum();
        println(sum + stones.size());
    }

    private static void part_2(List<String> lines) {
        var stones = Arrays.stream(lines.getFirst().split(" ")).mapToLong(Long::valueOf).boxed().toList();
        var sum = stones.stream().mapToLong(s -> blinkStone(s, 75, new HashMap<>())).sum();
        println(sum + stones.size());
    }

    private static Long blinkStone(Long stone, Integer times, Map<String, Long> cache) {

        var cached = cache.get(getKey(stone,times));
        if(cached != null){
            return cached;
        }

        if (times == 1) {
            return (long)blinkStone(stone).size() - 1;
        }
        var result = 0L;
        var stones = blinkStone(stone);
        result += blinkStone(stone).size() - 1;
        result += stones.stream().map(s -> blinkStone(s, times - 1, cache)).mapToLong(it -> it).sum();

        cache.put(getKey(stone, times), result);

        return result;
    }

    private static String getKey(Long stone, Integer times){
        return stone.toString() + "_" + times;
    }

    private static List<Long> blinkStone(Long stone) {
        if (stone == 0L) return Collections.singletonList(1L);
        var asString = stone.toString();

        if (asString.length() % 2 == 0) {
            var split = asString.split("");
            var left = Long.valueOf(Arrays.stream(split).limit(split.length / 2).collect(Collectors.joining("")));
            var right = Long.valueOf(Arrays.stream(split).skip(split.length / 2).collect(Collectors.joining("")));
            return List.of(left, right);
        } else {
            return Collections.singletonList(stone * 2024);
        }
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_11.kt
package org.twyszomirski.aoc.solutions

import java.io.File


class Day_11 {

    fun solve() {
        val lines = File("src/main/resources/input_day_11.txt").readLines()
        println("======== Day 11 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = input.first().split(" ").filter { it.isNotBlank() }.map { it.toLong() }
        val sum = data.sumOf { blinkStone(listOf(it), 25) }
        println(sum)
    }

    fun part2(input: List<String>) {
        val data = input.first().split(" ").filter { it.isNotBlank() }.map { it.toLong() }
        val cache = mutableMapOf<String, Long>()
        val sum = data.sumOf { blinkStoneWithImprovments(it, 75, cache) }
        println(sum + data.size)
    }


    fun blinkStone(stones: List<Long>, times: Int): Int {
        var current = stones
        repeat(times) {
            current = current.map { stone ->
                val stoneSize = stone.toString().length
                when {
                    stone == 0L -> listOf(1L)
                    stoneSize % 2 == 0 -> listOf(
                        stone.toString().substring(0, stoneSize / 2).toLong(),
                        stone.toString().substring(stoneSize / 2, stoneSize).toLong()
                    )

                    else -> listOf(stone * 2024)
                }
            }.flatten()
        }
        return current.size
    }


    fun blinkStoneWithImprovments(stone: Long, times: Int, cache: MutableMap<String, Long>): Long {

        if (times == 0) {
            return 0
        }

        val cacheKey = cacheKey(stone, times)
        val hit = cache[cacheKey]
        if (hit != null) {
            return hit
        }

        val stoneSize = stone.toString().length
        var local = 0
        val global = when {
            stone == 0L -> blinkStoneWithImprovments(1L, times - 1, cache)
            stoneSize % 2 == 0 -> {
                local = 1
                blinkStoneWithImprovments(
                    stone.toString().substring(0, stoneSize / 2).toLong(), times - 1, cache
                ) +
                        blinkStoneWithImprovments(
                            stone.toString().substring(stoneSize / 2, stoneSize).toLong(),
                            times - 1,
                            cache
                        )

            }

            else -> blinkStoneWithImprovments(stone * 2024, times - 1, cache)
        }
        val result = global + local
        cache.put(stone.toString() + "_" + times, result)
        return result
    }

    private fun cacheKey(stone: Long, times: Int) = stone.toString() + "_" + times
}
